ext/opentelemetry-ext-cloud-trace/tests/test_cloud_trace_exporter.py

```diff
@@ -0,0 +1,245 @@
+        )
+
+    def test_extract_attributes(self):
+        self.assertEqual(_extract_attributes({}), {"attribute_map": {}})
+        self.assertEqual(
+            _extract_attributes(self.attributes_variety_pack),
+            self.extracted_attributes_variety_pack,
+        )
+        # Test ignoring attributes with illegal value type
+        self.assertEqual(
+            _extract_attributes({"illegal_attribute_value": dict()}),
+            {"attribute_map": {}},
+        )
+
+    def test_extract_events(self):
+        self.assertIsNone(_extract_events([]))
+        time_in_ns = 1589919268850900051
+        time_in_ms_and_ns = {"seconds": 1589919268, "nanos": 850899968}
+        time_in_ns2 = 1589919438550020326
+        time_in_ms_and_ns2 = {"seconds": 1589919438, "nanos": 550020352}
+        event1 = Event(
+            name="event1",
+            attributes=self.attributes_variety_pack,
+            timestamp=time_in_ns,
+        )
+        event2 = Event(
+            name="event2",
+            attributes={"illegal_attr_value": dict()},
+            timestamp=time_in_ns2,
+        )
+        self.assertEqual(
+            _extract_events([event1, event2]),
+            {
+                "time_event": [
+                    {
+                        "time": time_in_ms_and_ns,
+                        "annotation": {
+                            "description": {
+                                "value": "event1",
+                                "truncated_byte_count": 0,
+                            },
+                            "attributes": self.extracted_attributes_variety_pack,
+                        },
+                    },
+                    {
+                        "time": time_in_ms_and_ns2,
+                        "annotation": {
+                            "description": {
+                                "value": "event2",
+                                "truncated_byte_count": 0,
+                            },
+                            "attributes": {"attribute_map": {}},
+                        },
+                    },
+                ]
+            },
+        )
+
+    def test_extract_links(self):
+        self.assertIsNone(_extract_links([]))
+        trace_id = "6e0c63257de34c92bf9efcd03927272e"
+        span_id = "95bb5edabd45950f"
+        link1 = Link(
+            context=SpanContext(
+                trace_id=int(trace_id, 16),
+                span_id=int(span_id, 16),
+                is_remote=False,
+            ),
+            attributes={},
+        )
+        link2 = Link(
+            context=SpanContext(
+                trace_id=int(trace_id, 16),
+                span_id=int(span_id, 16),
+                is_remote=False,
+            ),
+            attributes=self.attributes_variety_pack,
+        )
+        link3 = Link(
+            context=SpanContext(
+                trace_id=int(trace_id, 16),
+                span_id=int(span_id, 16),
```

Comment:
I assume these are links to other spans; would it be more realistic to give the links different span ids?

Reply:
good point, will do